src/FffGcodeWriter.cpp

```diff
@@ -1660,57 +1732,93 @@ bool FffGcodeWriter::addSupportInfillToGCode(const SliceDataStorage& storage, La
     EFillMethod support_pattern = infill_extr.getSettingAsFillMethod("support_pattern"); // first layer pattern must be same as other layers
     if (layer_nr <= 0 && (support_pattern == EFillMethod::LINES || support_pattern == EFillMethod::ZIG_ZAG)) { support_pattern = EFillMethod::GRID; }
 
-    int infill_extruder_nr_here = (layer_nr <= 0)? getSettingAsIndex("support_extruder_nr_layer_0") : getSettingAsIndex("support_infill_extruder_nr");
+    int infill_extruder_nr_here = (layer_nr <= 0) ? getSettingAsIndex("support_extruder_nr_layer_0") : getSettingAsIndex("support_infill_extruder_nr");
     const ExtruderTrain& infill_extr_here = *storage.meshgroup->getExtruderTrain(infill_extruder_nr_here);
 
-    const Polygons& support = support_layer.supportAreas;
-
-    std::vector<PolygonsPart> support_islands = support.splitIntoParts();
-
+    // create a list of outlines and use PathOrderOptimizer to optimize the travel move
+    std::vector<Polygons> infill_outline_list;
+    for (unsigned int i = 0; i < support_layer.support_infill_part_list.size(); ++i)
+    {
+        infill_outline_list.push_back(support_layer.support_infill_part_list[i].outline);
+    }
     PathOrderOptimizer island_order_optimizer(gcode_layer.getLastPosition());
-    for(unsigned int n=0; n<support_islands.size(); n++)
+    for (unsigned int i = 0; i < infill_outline_list.size(); ++i)
     {
-        island_order_optimizer.addPolygon(support_islands[n][0]);
+        island_order_optimizer.addPolygon(infill_outline_list[i][0]);
     }
     island_order_optimizer.optimize();
 
-    for(unsigned int n=0; n<support_islands.size(); n++)
+    // process each support infill islands
+    for (int support_infill_part_idx : island_order_optimizer.polyOrder)
```

Comment:
This function is quite different from `processSingleLayerInfill`.
Below I discuss some ways in which these differences cause problems.

If you managed to fix these problems I think this function is a lot more like `processSingleLayerInfill`.
You can then try to remove the code duplication by letting both of these functions call a function which has the code common to both these.

Let's first do the other refactors and then see whether it's feasible to remove the code duplication.